crates/pipeline_manager/src/db/storage.rs

```diff
@@ -363,54 +363,71 @@ pub(crate) trait Storage {
+    /// identifier or name.
     async fn new_service(
         &self,
         tenant_id: TenantId,
         id: Uuid,
         name: &str,
         description: &str,
         config: &ServiceConfig,
+        txn: Option<&Transaction<'_>>,
     ) -> Result<ServiceId, DBError>;
 
     /// Retrieves a list of all services of a tenant.
-    async fn list_services(&self, tenant_id: TenantId) -> Result<Vec<ServiceDescr>, DBError>;
+    /// Optionally, filtered by service configuration type.
+    async fn list_services(
+        &self,
+        tenant_id: TenantId,
+        filter_config_type: &Option<&str>,
+    ) -> Result<Vec<ServiceDescr>, DBError>;
 
-    /// Retrieves service descriptor for the given
-    /// `service_id`.
+    /// Retrieves service descriptor for the given `service_id`.
+    ///
+    /// Returns error if there does not exist a service with the provided
+    /// identifier.
     async fn get_service_by_id(
         &self,
         tenant_id: TenantId,
         service_id: ServiceId,
+        txn: Option<&Transaction<'_>>,
     ) -> Result<ServiceDescr, DBError>;
 
-    /// Retrieves service descriptor for the given unique `name`.
+    /// Retrieves service descriptor for the given unique service `name`.
+    ///
+    /// Returns error if there does not exist a service with the provided name.
     async fn get_service_by_name(
         &self,
         tenant_id: TenantId,
-        name: String,
+        name: &str,
+        txn: Option<&Transaction<'_>>,
     ) -> Result<ServiceDescr, DBError>;
 
-    /// Updates existing service.
-    /// Only the description and config can be updated, as the name is
-    /// immutable. The description must be provided and will always be updated.
-    /// The config is optional and not updated if not provided.
+    /// Updates the name, description and/or configuration of the existing
+    /// service, of which the identifier is provided. If new values are not
+    /// provided, the existing values in storage are kept.
+    ///
+    /// Returns error if there does not exist a service with the provided
+    /// identifier.
     async fn update_service(
         &self,
         tenant_id: TenantId,
         service_id: ServiceId,
-        description: &str,
+        name: &Option<&str>,
+        description: &Option<&str>,
         config: &Option<ServiceConfig>,
+        txn: Option<&Transaction<'_>>,
     ) -> Result<(), DBError>;
 
-    /// Deletes by id the service from the database.
-    /// TODO: what are pre-conditions for successful deletion?
-    /// TODO: what are post-conditions after successful deletion
-    ///       (e.g., cascading)?
-    async fn delete_service(
-        &self,
-        tenant_id: TenantId,
-        service_id: ServiceId,
-    ) -> Result<(), DBError>;
+    /// Deletes the service by its provided name.
+    ///
+    /// TODO: define pre-conditions (e.g., does it not need to be in use)
+    /// TODO: define post-conditions including cascading effect
```

Comment:
These cases should be handled for you by integrity constraints in the database. Just make sure to return the appropriate error types.

Reply:
My motivation here is that currently in this PR, there are not yet any constraints. In the following PR, connector will likely have a foreign key constraint to the service. The reason I keep these comments there is as reminder to think about what the cascading effect of the foreign key should be (CASCADE, SET NULL, or RESTRICT). Note this is different from pipeline to connector, as there is an attach table in between there.

Reply:
The default behavior like with other links in our APIs (e.g., pipeline -> program) is to prevent deletions if an object is "in use" (i.e., has a reference to it from some other object).

Reply:
Marking as resolved.